### PR TITLE
[DON'T MERGE] refactor: enforce required pegin confirmation depth by network

### PR DESCRIPTION
**Pull request by scottmillner**
tested running sidecar related pr as well and behaving as expected

pr https://github.com/botanix-labs/Macbeth/pull/115 supersedes this one

### CODE REVIEW

File: crates/consensus/authority/src/block_builder.rs [hunk 1]
```diff
@@ -159,10 +167,13 @@ where
         };
 
         // Process Botanix specific logs
+        let is_signet = self.chain_spec.chain().id() == BOTANIX_TESTNET.chain().id();
```
Comment (arminsabouri) [inline crates/consensus/authority/src/block_builder.rs]: nit: Do we you mind if we switch the this and other variable names to `testnet` to remain consistent with the other variables? i.e  `is_testnet`

Comment (scottmillner) [reply crates/consensus/authority/src/block_builder.rs]: you got it! done!

File: crates/consensus/authority/src/block_builder.rs [hunk 2]
```diff
@@ -136,6 +136,14 @@ where
             .unzip();
 
         let recent_bitcoin_block_header = *self.bitcoin_block_header.read().await;
+        let recent_bitcoin_block_height = match recent_bitcoin_block_header {
```
Comment (ozymandias-rex) [inline crates/consensus/authority/src/block_builder.rs]: given this is the same as https://github.com/botanix-labs/Macbeth/pull/110/files#diff-31fa1cff6a450309ffc5d47fb4caababd00c4d0f542a863e29d96e9c284c59eeR114, does it make sense to convert it to a helper in utils and import it into  the block builder and block fetcher

File: crates/primitives/src/constants/mod.rs [hunk 3]
```diff
@@ -192,6 +192,12 @@ pub const STAKING_CONTRACT_ADDRESS: &str = "0xA107e8549eB97f420a1D3b93Bd202e0845
 /// Storage slot index of list of stakers
 pub const STAKER_BALANCE_MAPPING_STORAGE_SLOT_INDEX: u64 = 1u64;
 
+/// Signet required confirmation depth for pegins
+pub const SIGNET_PEGIN_CONFIRMATION_DEPTH: u32 = 1;
+
+/// Mainnet required confirmation depth for pegins
+pub const MAINNET_PEGIN_CONFIRMATION_DEPTH: u32 = 100;
```
Comment (ozymandias-rex) [inline crates/primitives/src/constants/mod.rs:201]: Does this translate to 100 confirmations for mainnet ? How do we arrive as it ? Is there a property of pegins that necessitates the high threshold ?

File: crates/consensus/authority/src/utils.rs [hunk 4]
```diff
@@ -170,6 +189,16 @@ async fn process_botanix_log(
                 info!(target: "consensus::authority", "Parsing and sending minting event to btc_server");
                 let pegin_data = parse_pegin_reth_log_topic(log)
                     .expect("passed evm check should pass this parse attempt");
+                // enforce required confirmation depth by network
+                let confirmation_depth = if is_testnet {
```
Comment (ozymandias-rex) [inline crates/consensus/authority/src/utils.rs]: nit: using a `match` here might be more idiomatic

```rust
let confirmation_depth = match is_testnet {
    true => SIGNET_PEGIN_CONFIRMATION_DEPTH,
    false => MAINNET_PEGIN_CONFIRMATION_DEPTH,
};
```

File: crates/consensus/authority/src/utils.rs [hunk 5]
```diff
@@ -170,6 +189,16 @@ async fn process_botanix_log(
...
+                    SIGNET_PEGIN_CONFIRMATION_DEPTH
+                } else {
+                    MAINNET_PEGIN_CONFIRMATION_DEPTH
+                };
+                if pegin_data.bitcoin_block_height > recent_bitcoin_block_height - confirmation_depth {
+                    info!(target: "consensus::authority", "pegin confirmation depth not met, skipping");
```
Comment (ozymandias-rex) [inline crates/consensus/authority/src/utils.rs]: Should this be a `warn!` ?

File: crates/consensus/authority/src/block_builder.rs (hunk 2 above)
Comment (scottmillner) [reply crates/consensus/authority/src/block_builder.rs]: i certainly can
my general rule is if i have to do it 3 times i pull it out into a separate function lol
i shall change the rule to 2 times and convert it to a helper!

implemented

File: crates/primitives/src/constants/mod.rs (hunk 3 above)
Comment (scottmillner) [reply crates/primitives/src/constants/mod.rs:201]: btc protocol doesn't allow reorgs after 100 blocks
 @0xBEEFCAF3 correct?

File: crates/consensus/authority/src/utils.rs (hunk 4 above)
Comment (scottmillner) [reply crates/consensus/authority/src/utils.rs]: thanks, implemented!

File: crates/primitives/src/constants/mod.rs (hunk 3 above)
Comment (ozymandias-rex) [reply crates/primitives/src/constants/mod.rs:201]: IIRC , the current code will skip peggins until they have reached 100 confirmations (600 mins). This is alot more than the time that exchanges wait for .

File: crates/consensus/authority/src/utils.rs (hunk 5 above)
Comment (scottmillner) [reply crates/consensus/authority/src/utils.rs]: yes it should, implemented, thanks!

File: crates/primitives/src/constants/mod.rs (hunk 3 above)
Comment (scottmillner) [reply crates/primitives/src/constants/mod.rs:201]: @ozymandias-rex 
We're trying to ensure that a reorg is never possible and 100 is the only depth where that is guaranteed.
Currently, a depth of 38 will guarantee that with the current hashrate (see [calculator](https://jlopp.github.io/bitcoin-confirmation-risk-calculator/) but this is a moving target). 100 is future proof unless a protocol change is made.

But agreed, this is a major constraint and may not be practical so something we need to discuss as a team.

Comment (arminsabouri) [reply crates/primitives/src/constants/mod.rs:201]: As it turns out there isnt a protocol limit. 100 is just what liquid uses to reduce the probability of a re-org to pretty much 0

File: crates/consensus/authority/src/utils.rs [hunk 6]
```diff
@@ -250,11 +286,31 @@ pub(crate) fn is_epoch_end(current_block_number: u64) -> bool {
...
+///
+/// # Arguments
+///
+/// * `recent_bitcoin_block_header` - The recent bitcoin block header
+///
+/// # Returns
+///
+/// Returns the recent block height or 0 if None.
+pub(crate) fn get_recent_block_height_or_zero(recent_bitcoin_block_header: Option<(Header, u32)>) -> u32 {
+    match recent_bitcoin_block_header {
```
Comment (BatiGencho) [inline crates/consensus/authority/src/utils.rs]: nit: I believe we could rewrite this in a more rusty way by:
```
recent_bitcoin_block_header.map(|header| header.height).unwrap_or_else(|| {
   error!(target: "consensus::authority", "Failed to get recent bitcoin block height");
            0
})
```

File: crates/consensus/authority/src/utils.rs [hunk 7]
```diff
@@ -170,6 +195,17 @@ async fn process_botanix_log(
                 info!(target: "consensus::authority", "Parsing and sending minting event to btc_server");
                 let pegin_data = parse_pegin_reth_log_topic(log)
                     .expect("passed evm check should pass this parse attempt");
+                // enforce required confirmation depth by network
+                let confirmation_depth = match is_testnet {
+                    true => SIGNET_PEGIN_CONFIRMATION_DEPTH,
+                    false => MAINNET_PEGIN_CONFIRMATION_DEPTH,
+                };
```
Comment (BatiGencho) [inline crates/consensus/authority/src/utils.rs]: I would personally move this into its own function e.g. 
```
fn get_confirmation_depth(is_testnet) {
 match is_testnet {
                    true => SIGNET_PEGIN_CONFIRMATION_DEPTH,
                    false => MAINNET_PEGIN_CONFIRMATION_DEPTH,
                }
}
```
as we could later also evtl. extend this with regtest network and others ? Maybe not :)

File: crates/consensus/authority/src/block_fetcher.rs [hunk 8]
```diff
@@ -126,10 +131,13 @@ where
                         SealedBlockWithSenders::new(sealed_block.clone(), senders)
                             .expect("senders are valid");
                     // Process Botanix specific logs
+                    let is_testnet = self.chain_spec.chain().id() == BOTANIX_TESTNET.chain().id();
```
Comment (BatiGencho) [inline crates/consensus/authority/src/block_fetcher.rs]: I guess we could also pull this into its own function for later re-usability

e,g.
```
fn is_testnet() -> bool {
self.chain_spec.chain().id() == BOTANIX_TESTNET.chain().id()
}
```

File: crates/consensus/authority/src/block_builder.rs [hunk 9]
```diff
@@ -159,10 +168,13 @@ where
         };
 
         // Process Botanix specific logs
+        let is_testnet = self.chain_spec.chain().id() == BOTANIX_TESTNET.chain().id();
```
Comment (BatiGencho) [inline crates/consensus/authority/src/block_builder.rs]: same here, see my comment below reg. reusability

File: crates/primitives/src/constants/mod.rs [hunk 10]
```diff
@@ -192,6 +192,12 @@ pub const STAKING_CONTRACT_ADDRESS: &str = "0xA107e8549eB97f420a1D3b93Bd202e0845
 /// Storage slot index of list of stakers
 pub const STAKER_BALANCE_MAPPING_STORAGE_SLOT_INDEX: u64 = 1u64;
 
+/// Signet required confirmation depth for pegins
+pub const SIGNET_PEGIN_CONFIRMATION_DEPTH: u32 = 1;
```
Comment (BatiGencho) [inline crates/primitives/src/constants/mod.rs:198]: Hmm... is this something that should live among the reth primitives or can we not move it into some of our botanix poa specific rust files ? Ignore if comment is stupid pls

File: crates/consensus/authority/src/utils.rs (hunk 6 above)
Comment (scottmillner) [reply crates/consensus/authority/src/utils.rs]: that's better for sure, thanks!
done

File: crates/consensus/authority/src/utils.rs (hunk 7 above)
Comment (scottmillner) [reply crates/consensus/authority/src/utils.rs]: you got it!
done!

File: crates/primitives/src/constants/mod.rs (hunk 10 above)
Comment (scottmillner) [reply crates/primitives/src/constants/mod.rs:198]: added a TODO
if we move these we should move all the botanix specific consts in this file(`STAKING_CONTRACT_ADDRESS` and `STAKER_BALANCE_MAPPING_STORAGE_SLOT_INDEX`)

maybe add a new file in this folder (constants)?

File: crates/consensus/authority/src/block_fetcher.rs (hunk 8 above)
Comment (scottmillner) [reply crates/consensus/authority/src/block_fetcher.rs]: done!

File: crates/consensus/authority/src/block_builder.rs (hunk 9 above)
Comment (scottmillner) [reply crates/consensus/authority/src/block_builder.rs]: done!

File: crates/primitives/src/constants/mod.rs (hunk 10 above)
Comment (BatiGencho) [reply crates/primitives/src/constants/mod.rs:198]: Nah, its fine. Now that I think about it, leave them in there, we have already massively protruded the reth repo with our impls, why not to keep botanix and reth consts in the same place.

**Review by BatiGencho — APPROVED** [review]

**Comment by scottmillner** [comment]
implemented in pr https://github.com/botanix-labs/Macbeth/pull/115

**Pull request closed without merging**